File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', '}': '{', ']': '['}
        for char in s:
            if char in mapping:                                                 # if it's a closing bracket
                top = stack.pop() if stack else "#"                             # pop or dummy
                if mapping[char] != top: return False
            else:
                stack.append(char)
        return len(stack) == 0

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        if (s.length() % 2 != 0) {
            return false;
        }
        char[] par = new char[s.length()];
        int top = -1;
        for (char c : s.toCharArray()) {
            if(c == '{'){
                par[++top] = '}';
            }else if(c == '['){
                par[++top] = ']';
            }else if(c == '('){
                par[++top] = ')';
            }else if(top == -1 || par[top--] != c){
                return false;
            }
        }
        return top == -1;
    }
}